lxc/completion.go

```diff
@@ -250,44 +245,144 @@ func (g *cmdGlobal) cmpInstanceKeys(instanceName string) ([]string, cobra.ShellC
+	if err != nil {
+		return nil, cobra.ShellCompDirectiveError
+	}
+
+	instanceConfig, ok := metadataConfiguration.Configs["instance"]
+	if !ok {
+		return nil, cobra.ShellCompDirectiveError
 	}
 
-	for k := range instancetype.InstanceConfigKeysAny {
-		keys = append(keys, k)
+	// Pre-allocate configKeys slice capacity.
+	keyCount := 0
+	for _, field := range instanceConfig {
+		keyCount += len(field.Keys)
 	}
 
-	return keys, cmpDirectives
+	configKeys := make([]string, 0, keyCount)
+
+	for _, field := range instanceConfig {
+		for _, key := range field.Keys {
+			for configKey, configKeyField := range key {
+				if strings.HasSuffix(configKey, "*") {
+					configKey = strings.Trim(configKey, "*")
+				}
+
+				if instanceType == string(api.InstanceTypeContainer) && configKeyField.Condition == "container" {
+					configKeys = append(configKeys, configKey)
+				} else if instanceType == string(api.InstanceTypeVM) && configKeyField.Condition == "virtual machine" {
+					configKeys = append(configKeys, configKey)
+				}
+
+				// InstanceTypeAny config keys.
+				if configKeyField.Condition == "" {
+					configKeys = append(configKeys, configKey)
+				}
+			}
+		}
+	}
+
+	return configKeys, cmpDirectives | cobra.ShellCompDirectiveNoSpace
 }
 
 // cmpInstanceAllKeys provides shell completion for all possible instance configuration keys.
 // It returns a list of all possible instance configuration keys along with a shell completion directive.
-func (g *cmdGlobal) cmpInstanceAllKeys() ([]string, cobra.ShellCompDirective) {
-	keys := make([]string, 0, len(instancetype.InstanceConfigKeysContainer)+len(instancetype.InstanceConfigKeysVM)+len(instancetype.InstanceConfigKeysAny))
+func (g *cmdGlobal) cmpInstanceAllKeys(profileName string) ([]string, cobra.ShellCompDirective) {
 	cmpDirectives := cobra.ShellCompDirectiveNoFileComp
 
-	for k := range instancetype.InstanceConfigKeysContainer {
-		keys = append(keys, k)
+	// Parse remote
+	resources, err := g.ParseServers(profileName)
+	if err != nil {
+		return nil, cobra.ShellCompDirectiveError
 	}
 
-	for k := range instancetype.InstanceConfigKeysVM {
-		keys = append(keys, k)
+	resource := resources[0]
+	client := resource.server
+
+	metadataConfiguration, err := client.GetMetadataConfiguration()
+	if err != nil {
+		return nil, cobra.ShellCompDirectiveError
 	}
 
-	for k := range instancetype.InstanceConfigKeysAny {
-		keys = append(keys, k)
+	instanceConfig, ok := metadataConfiguration.Configs["instance"]
+	if !ok {
+		return nil, cobra.ShellCompDirectiveError
 	}
 
-	return keys, cmpDirectives
+	// Pre-allocate configKeys slice capacity.
+	keyCount := 0
+	for _, field := range instanceConfig {
+		keyCount += len(field.Keys)
+	}
+
+	configKeys := make([]string, 0, keyCount)
+
+	for _, field := range instanceConfig {
+		for _, key := range field.Keys {
+			for configKey := range key {
+				if strings.HasSuffix(configKey, "*") {
+					configKey = strings.Trim(configKey, "*")
```

Comment:
`strings.TrimSuffix()` is a faster implementation than .Trim()`.

`configKey = configKey[:len(configKey)-1]` would also work but might be less readable?

Reply:
I say let's go with the slice operation for performance :)

Reply:
Since it's just `len(configKey)-1`, I think readability is intact with the slicing syntax.

Reply:
Hmm my language server is suggesting `strings.TrimSuffix()` over the conditional in conjunction with slicing syntax. Since `strings.TrimSuffix()` returns the string unchanged and has early exits, the overhead may be negligible if we drop the conditional.